File: bindings/python/examples/error_handling.py
#!/usr/bin/env python3
"""
Error Handling Best Practices

This example demonstrates:
1. Proper exception handling
2. Timeout handling
3. Validation errors
4. Retry strategies
"""

import logging
import os
import time

from kaleidoswap import KaleidoClient, KaleidoConfig, KaleidoError

# Configure logging
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger(__name__)

# Configuration
API_URL = os.getenv("KALEIDO_API_URL", "https://api.staging.kaleidoswap.com")


def example_basic_error_handling():
    """Demonstrate basic error handling."""
    logger.info("\n🔐 Example 1: Basic Error Handling")

    config = KaleidoConfig(base_url=API_URL)
    client = KaleidoClient(config)

    try:
        assets = client.list_assets()
        logger.info(f"  ✅ Found {len(assets)} assets")
    except KaleidoError as e:
        logger.warning(f"  ⚠️  Kaleido API error: {e}")
    except Exception as e:
        logger.error(f"  ❌ Unexpected error: {type(e).__name__}: {e}")


def example_invalid_parameters():
    """Demonstrate handling invalid parameters."""
    logger.info("\n✅ Example 2: Invalid Parameter Handling")

    config = KaleidoConfig(base_url=API_URL)
    client = KaleidoClient(config)

    try:
        # Try to get a swap status with invalid hash
        client.get_swap_status("invalid_payment_hash_12345")
        logger.info("  Got swap status (unexpected)")
    except KaleidoError as e:
        logger.warning(f"  ⚠️  API error (expected): {e}")
    except Exception as e:
        logger.info(f"  Got error: {type(e).__name__}: {e}")


def example_validation():
    """Demonstrate input validation."""
    logger.info("\n⚠️  Example 3: Input Validation")

    config = KaleidoConfig(base_url=API_URL)
    client = KaleidoClient(config)

    # Validate negative amount
    logger.info("  Testing negative amount...")
    result = client.validate_amount(-100, "BTC")
    if not result["valid"]:
        logger.info(f"    ❌ Invalid: {result['errors']}")
    else:
        logger.info("    ✅ Valid (unexpected)")

    # Validate zero amount
    logger.info("  Testing zero amount...")
    result = client.validate_amount(0, "BTC")
    if not result["valid"]:
        logger.info(f"    ❌ Invalid: {result['errors']}")
    else:
        logger.info("    ✅ Valid (unexpected)")

    # Validate valid amount
    logger.info("  Testing valid amount (0.01 BTC)...")
    try:
        result = client.validate_amount(0.01, "BTC")
        if result["valid"]:
            logger.info(f"    ✅ Valid: {result['raw_amount']} atomic units")
        else:
            logger.info(f"    ❌ Invalid: {result['errors']}")
    except Exception as e:
        logger.info(f"    Could not validate: {e}")


def example_retry_strategy():
    """Demonstrate retry strategy for transient errors."""
    logger.info("\n🔄 Example 4: Retry Strategy")

    config = KaleidoConfig(base_url=API_URL)
    client = KaleidoClient(config)

    max_retries = 3
    retry_delay = 1.0

    for attempt in range(max_retries):
        try:
            logger.info(f"  Attempt {attempt + 1}/{max_retries}...")
            assets = client.list_assets()
            logger.info(f"  ✅ Success! Found {len(assets)} assets")
            break
        except KaleidoError as e:
            if attempt < max_retries - 1:
                logger.warning(f"  ⚠️  API error, retrying in {retry_delay}s: {e}")
                time.sleep(retry_delay)
                retry_delay *= 2  # Exponential backoff
            else:
                logger.error(f"  ❌ Failed after {max_retries} attempts: {e}")
        except Exception as e:
            logger.error(f"  ❌ Unexpected error: {e}")
            break


def example_invalid_pair_ticker():
    """Demonstrate handling invalid pair ticker."""
    logger.info("\n📊 Example 5: Invalid Pair Ticker")

    config = KaleidoConfig(
        base_url=API_URL,
        node_url="http://localhost:3001",
    )
    client = KaleidoClient(config)

    try:
        # Invalid pair ticker format (missing slash)
        client.execute_swap_by_pair(
            pair_ticker="BTCUSDT",  # Should be BTC/USDT
            amount=0.01,
            side="sell",
            taker_pubkey="fakepubkey",
        )
    except ValueError as e:
        logger.info(f"  ⚠️  ValueError (expected): {e}")
    except Exception as e:
        logger.info(f"  Got error: {type(e).__name__}: {e}")

    try:
        # Invalid side
        client.execute_swap_by_pair(
            pair_ticker="BTC/USDT",
            amount=0.01,
            side="hodl",  # Invalid side
            taker_pubkey="fakepubkey",
        )
    except ValueError as e:
        logger.info(f"  ⚠️  ValueError for invalid side (expected): {e}")
    except Exception as e:
        logger.info(f"  Got error: {type(e).__name__}: {e}")


def example_graceful_degradation():
    """Demonstrate graceful degradation when services are unavailable."""
    logger.info("\n🛡️  Example 6: Graceful Degradation")

    config = KaleidoConfig(base_url=API_URL)
    client = KaleidoClient(config)

    # Check if node is available
    if client.has_node():
        logger.info("  ✅ Node is configured")
        try:
            info = client.get_rgb_node_info()
            logger.info(f"    Node info: {info[:100]}...")
        except Exception as e:
            logger.warning(f"    Could not get node info: {e}")
    else:
        logger.info("  ℹ️  No node configured - operating in API-only mode")
        logger.info("    Some features like whitelist_trade won't be available")

    # Node operations will fail gracefully
    if not client.has_node():
        logger.info("  Skipping node-dependent operations")
    

def main():
    """Run all error handling examples."""
    logger.info("=" * 60)
    logger.info("Error Handling Best Practices Examples")
    logger.info("=" * 60)

    try:
        example_basic_error_handling()
        example_invalid_parameters()
        example_validation()
        example_retry_strategy()
        example_invalid_pair_ticker()
        example_graceful_degradation()

        logger.info("\n" + "=" * 60)
        logger.info("✅ All error handling examples completed!")
        logger.info("=" * 60)

    except Exception as e:
        logger.error(f"❌ Unexpected error in main: {e}", exc_info=True)


if __name__ == "__main__":
    main()

File: bindings/typescript/typescript/examples/error-handling.ts
/**
 * Error Handling Best Practices
 *
 * This example demonstrates:
 * 1. Proper exception handling with typed errors
 * 2. Retry strategies
 * 3. Graceful degradation
 * 4. Working with the error hierarchy
 */

import {
    KaleidoClient,
    KaleidoError,
    APIError,
    NetworkError,
    ValidationError,
    NotFoundError,
    ConfigError,
    type KaleidoConfig,
} from '../src/index.js';

const API_URL = process.env.KALEIDO_API_URL || 'https://api.regtest.kaleidoswap.com';

// ============================================================================
// Example 1: Basic Error Handling with Typed Exceptions
// ============================================================================

async function exampleTypedErrorHandling() {
    console.log('\n🔐 Example 1: Typed Error Handling');
    console.log('-'.repeat(50));

    const client = await KaleidoClient.create({ baseUrl: API_URL });

    try {
        // Try to get a non-existent asset
        const asset = await client.getAssetByTicker('NONEXISTENT_ASSET');
        console.log(`  Found asset: ${asset.name}`);
    } catch (e) {
        if (e instanceof NotFoundError) {
            console.log(`  ℹ️  Asset not found: ${e.message}`);
        } else if (e instanceof APIError) {
            console.log(`  ⚠️  API error (${e.statusCode}): ${e.message}`);
        } else if (e instanceof NetworkError) {
            console.log(`  🔌 Network error: ${e.message}`);
        } else if (e instanceof KaleidoError) {
            console.log(`  ❌ Kaleido error: ${e.message}`);
        } else {
            console.log(`  ❌ Unexpected error: ${e}`);
        }
    }
}

// ============================================================================
// Example 2: Error Hierarchy Navigation
// ============================================================================

async function exampleErrorHierarchy() {
    console.log('\n📋 Example 2: Error Hierarchy');
    console.log('-'.repeat(50));

    const client = await KaleidoClient.create({ baseUrl: 'https://invalid-url-that-does-not-exist.com' });

    try {
        await client.listAssets();
    } catch (e) {
        // All SDK errors extend KaleidoError
        if (e instanceof KaleidoError) {
            console.log(`  Error type: ${e.constructor.name}`);
            console.log(`  Message: ${e.message}`);

            // Check specific error types
            if (e instanceof NetworkError) {
                console.log(`  ↳ This is a network-related error`);
            }
        }
    }
}

// ============================================================================
// Example 3: Retry Strategy with Exponential Backoff
// ============================================================================

async function exampleRetryStrategy() {
    console.log('\n🔄 Example 3: Retry Strategy');
    console.log('-'.repeat(50));

    const maxRetries = 3;
    let retryDelay = 1000;

    const client = await KaleidoClient.create({ baseUrl: API_URL });

    for (let attempt = 1; attempt <= maxRetries; attempt++) {
        try {
            console.log(`  Attempt ${attempt}/${maxRetries}...`);
            const assets = await client.listAssets();
            console.log(`  ✅ Success! Found ${assets.length} assets`);
            return;
        } catch (e) {
            // Only retry on network errors
            if (e instanceof NetworkError && attempt < maxRetries) {
                console.log(`  ⚠️  Network error, retrying in ${retryDelay}ms...`);
                await new Promise(resolve => setTimeout(resolve, retryDelay));
                retryDelay *= 2; // Exponential backoff
            } else if (e instanceof APIError && e.statusCode === 429) {
                // Rate limited - wait longer
                console.log(`  ⏳ Rate limited, waiting 5 seconds...`);
                await new Promise(resolve => setTimeout(resolve, 5000));
            } else {
                // Non-retryable error
                console.log(`  ❌ Non-retryable error: ${e}`);
                return;
            }
        }
    }

    console.log(`  ❌ Failed after ${maxRetries} attempts`);
}

// ============================================================================
// Example 4: Graceful Degradation
// ============================================================================

async function exampleGracefulDegradation() {
    console.log('\n🛡️  Example 4: Graceful Degradation');
    console.log('-'.repeat(50));

    // Client without RGB node
    const clientNoNode = await KaleidoClient.create({ baseUrl: API_URL });

    if (clientNoNode.hasNode()) {
        console.log('  ✓ RGB node is configured');
    } else {
        console.log('  ℹ️  No RGB node configured - operating in API-only mode');
        console.log('    Available: Market data, quotes');
        console.log('    Not available: Direct node operations');
    }

    // Client with RGB node
    const clientWithNode = await KaleidoClient.create({
        baseUrl: API_URL,
        nodeUrl: 'http://localhost:3001',
    });

    if (clientWithNode.hasNode()) {
        console.log('  ✓ RGB node is configured (with nodeUrl)');
    }
}

// ============================================================================
// Example 5: Validation Error Details
// ============================================================================

async function exampleValidationErrors() {
    console.log('\n✅ Example 5: Validation Errors');
    console.log('-'.repeat(50));

    const client = await KaleidoClient.create({ baseUrl: API_URL });

    try {
        // Try to get quote with invalid parameters
        await client.getQuote(
            'INVALID/PAIR',
            null,
            null,  // Both null - should fail validation
            'BTC_LN',
            'RGB_LN'
        );
    } catch (e) {
        if (e instanceof ValidationError) {
            console.log(`  Validation failed: ${e.message}`);
        } else {
            console.log(`  Error: ${e}`);
        }
    }
}

// ============================================================================
// Main
// ============================================================================

async function main() {
    console.log('='.repeat(60));
    console.log('Kaleidoswap SDK - Error Handling Examples');
    console.log('='.repeat(60));

    try {
        await exampleTypedErrorHandling();
        await exampleErrorHierarchy();
        await exampleRetryStrategy();
        await exampleGracefulDegradation();
        await exampleValidationErrors();

        console.log('\n' + '='.repeat(60));
        console.log('✅ All error handling examples completed!');
        console.log('='.repeat(60));
    } catch (e) {
        console.error('❌ Unexpected error in main:', e);
    }
}

main();
